File: BE_JAVA/src/main/java/com/lighting/service/AccessService.java
package com.lighting.service;

import java.util.Random;
import java.util.regex.Pattern;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.lighting.dto.AccountDto;
import com.lighting.entity.Account;
import com.lighting.repository.AccountRepository;
import com.lighting.util.EmailComposite;
import com.lighting.util.EmailService;
import com.lighting.util.JwtTokenUtil;
import com.lighting.util.RedisService;
import com.lighting.util.VerificationEmail;
import com.lighting.util.WelcomeEmail;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
@Transactional
public class AccessService {

    private static final Logger log = LoggerFactory.getLogger(AccessService.class);
    private static final Pattern EMAIL_PATTERN = Pattern.compile("^[A-Za-z0-9+_.-]+@(.+)$");

    private final AccountRepository accountRepository;
    private final EmailService emailService;
    private final RedisService redisService;
    private final JwtTokenUtil jwtTokenUtil;
    private final PasswordEncoder passwordEncoder;

    public void signUp(AccountDto.SignUpRequest request) {
        log.info("Sign up attempt for email: {}", request.getEmail());

        String emailSentKey = redisService.pinVerifyKey(request.getEmail()) + ":sent";
        if (redisService.get(emailSentKey) != null) {
            throw new RuntimeException("Verification email already sent. Please check your inbox.");
        }

        if (!EMAIL_PATTERN.matcher(request.getEmail()).matches()) {
            throw new RuntimeException("Invalid email");
        }

        if (accountRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Email already exists");
        }

        if (accountRepository.existsByPhone(request.getPhone())) {
            throw new RuntimeException("Phone already exists");
        }

        String code = getRandomSixDigit();

        // Use EmailComposite to send multiple emails
        EmailComposite emailComposite = new EmailComposite(request.getEmail());
        emailComposite.addComponent(new VerificationEmail(request.getEmail(), code));
        emailComposite.addComponent(new WelcomeEmail(request.getEmail()));
        emailService.sendEmailComponent(emailComposite);

        redisService.set(redisService.pinVerifyKey(request.getEmail()), code, 60);
        redisService.set(emailSentKey, "true", 60);

        redisService.set(redisService.pinVerifyKey(request.getEmail()) + ":firstName", request.getFirstName(), 6000);
        redisService.set(redisService.pinVerifyKey(request.getEmail()) + ":lastName", request.getLastName(), 6000);
        redisService.set(redisService.pinVerifyKey(request.getEmail()) + ":password", passwordEncoder.encode(request.getPassword()), 6000);
        redisService.set(redisService.pinVerifyKey(request.getEmail()) + ":phone", request.getPhone(), 6000);
        redisService.set(redisService.pinVerifyKey(request.getEmail()) + ":gender", request.getGender(), 6000);
        redisService.set(redisService.pinVerifyKey(request.getEmail()) + ":addressLine1", request.getAddressLine1(), 6000);
        redisService.set(redisService.pinVerifyKey(request.getEmail()) + ":province", request.getProvince(), 6000);
        redisService.set(redisService.pinVerifyKey(request.getEmail()) + ":district", request.getDistrict(), 6000);
        redisService.set(redisService.pinVerifyKey(request.getEmail()) + ":ward", request.getWard(), 6000);
        redisService.set(redisService.pinVerifyKey(request.getEmail()) + ":postalCode", request.getPostalCode(), 6000);
        redisService.set(redisService.pinVerifyKey(request.getEmail()) + ":addressNote", request.getAddressNote(), 6000);
        redisService.set(redisService.pinVerifyKey(request.getEmail()) + ":defaultAddress", String.valueOf(request.getDefaultAddress()), 6000);

        log.info("Verification and welcome emails sent to: {}", request.getEmail());
    }

    public void resetPassword(AccountDto.ResetPasswordRequest request) {
        log.info("Reset password attempt for email: {}", request.getEmail());

        String code = getRandomSixDigit();
        String subject = "Reset Password";
        String text = "Your code is: " + code;

        emailService.sendEmail(subject, text, request.getEmail());

        redisService.set(redisService.pinVerifyKey(request.getEmail()), code, 60);

        log.info("Reset password email sent to: {}", request.getEmail());
    }
    

    public Account verifyEmail(AccountDto.VerifyEmailRequest request) {
        log.info("Email verification attempt for: {}", request.getEmail());

        String storedCode = redisService.get(redisService.pinVerifyKey(request.getEmail()));

        if (storedCode == null || !storedCode.equals(request.getCode())) {
            throw new RuntimeException("Invalid code");
        }

        // Delete verification code from Redis
        redisService.delete(redisService.pinVerifyKey(request.getEmail()));
        
        // Get stored user information from Redis
        String storedPassword = redisService.get(redisService.pinVerifyKey(request.getEmail()) + ":password");
        String firstName = redisService.get(redisService.pinVerifyKey(request.getEmail()) + ":firstName");
        String lastName = redisService.get(redisService.pinVerifyKey(request.getEmail()) + ":lastName");
        String phone = redisService.get(redisService.pinVerifyKey(request.getEmail()) + ":phone");
        String gender = redisService.get(redisService.pinVerifyKey(request.getEmail()) + ":gender");
        String addressLine1 = redisService.get(redisService.pinVerifyKey(request.getEmail()) + ":addressLine1");
        String province = redisService.get(redisService.pinVerifyKey(request.getEmail()) + ":province");
        String district = redisService.get(redisService.pinVerifyKey(request.getEmail()) + ":district");
        String ward = redisService.get(redisService.pinVerifyKey(request.getEmail()) + ":ward");
        String postalCode = redisService.get(redisService.pinVerifyKey(request.getEmail()) + ":postalCode");
        String addressNote = redisService.get(redisService.pinVerifyKey(request.getEmail()) + ":addressNote");
        String defaultAddressStr = redisService.get(redisService.pinVerifyKey(request.getEmail()) + ":defaultAddress");
        
        if (storedPassword == null || firstName == null || lastName == null || phone == null) {
            throw new RuntimeException("Registration session expired. Please sign up again.");
        }
        
        // Delete all stored user data from Redis
        redisService.delete(redisService.pinVerifyKey(request.getEmail()) + ":password");
        redisService.delete(redisService.pinVerifyKey(request.getEmail()) + ":firstName");
        redisService.delete(redisService.pinVerifyKey(request.getEmail()) + ":lastName");
        redisService.delete(redisService.pinVerifyKey(request.getEmail()) + ":phone");
        redisService.delete(redisService.pinVerifyKey(request.getEmail()) + ":gender");
        redisService.delete(redisService.pinVerifyKey(request.getEmail()) + ":addressLine1");
        redisService.delete(redisService.pinVerifyKey(request.getEmail()) + ":province");
        redisService.delete(redisService.pinVerifyKey(request.getEmail()) + ":district");
        redisService.delete(redisService.pinVerifyKey(request.getEmail()) + ":ward");
        redisService.delete(redisService.pinVerifyKey(request.getEmail()) + ":postalCode");
        redisService.delete(redisService.pinVerifyKey(request.getEmail()) + ":addressNote");
        redisService.delete(redisService.pinVerifyKey(request.getEmail()) + ":defaultAddress");

        // Create new account with all user information
        Account account = new Account();
        account.setEmail(request.getEmail());
        account.setPassword(passwordEncoder.encode(storedPassword));
        account.setFirstName(firstName);
        account.setLastName(lastName);
        account.setPhone(phone);
        
        // Set optional fields if they exist
        if (gender != null && !gender.isEmpty()) {
            account.setGender(Account.Gender.valueOf(gender.toUpperCase()));
        }
        account.setAddressLine1(addressLine1);
        account.setProvince(province);
        account.setDistrict(district);
        account.setWard(ward);
        account.setPostalCode(postalCode);
        account.setAddressNote(addressNote);
        account.setDefaultAddress("true".equals(defaultAddressStr));
        
        // Set default values for other fields
        account.setRole(Account.Role.USER);
        account.setIsBlocked(false);

        Account savedAccount = accountRepository.save(account);

        log.info("Email verified and account created for: {}", request.getEmail());
        return savedAccount;
    }

    public AccountDto.AuthResponse signIn(AccountDto.SignInRequest request) {
        log.info("Sign in attempt for email: {}", request.getEmail());

        Account account = accountRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("Invalid username or password"));

        if (!passwordEncoder.matches(request.getPassword(), account.getPassword())) {
            throw new RuntimeException("Invalid username or password");
        }

        // Generate tokens
        String accessToken = jwtTokenUtil.generateAccessToken(
                account.getId(),
                account.getEmail(),
                account.getRole().getValue()
        );
        String refreshToken = jwtTokenUtil.generateRefreshToken(
                account.getId(),
                account.getEmail(),
                account.getRole().getValue()
        );

        // Store refresh token in Redis
        String refreshKey = redisService.refreshTokenKey(account.getId());
        redisService.delete(refreshKey); // Delete existing token
        redisService.set(refreshKey, refreshToken);

        // Create response
        AccountDto accountDto = mapToDto(account);
        AccountDto.AuthResponse response = new AccountDto.AuthResponse();
        response.setAccessToken(accessToken);
        response.setRefreshToken(refreshToken);
        response.setUser(accountDto);

        log.info("Sign in successful for: {}", request.getEmail());
        return response;
    }

    public AccountDto.AuthResponse refreshToken(AccountDto.RefreshTokenRequest request) {
        log.info("Refresh token attempt");

        String refreshToken = request.getRefreshToken();

        if (refreshToken == null || refreshToken.isEmpty()) {
            throw new RuntimeException("Unauthorized");
        }

        // Validate token
        if (!jwtTokenUtil.validateToken(refreshToken)) {
            throw new RuntimeException("Invalid refresh token");
        }

        if (jwtTokenUtil.isTokenExpired(refreshToken)) {
            throw new RuntimeException("Refresh token expired");
        }

        String userId = jwtTokenUtil.getUserIdFromToken(refreshToken);
        Account account = accountRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        // Verify stored refresh token
        String storedToken = redisService.get(redisService.refreshTokenKey(userId));
        if (storedToken == null || !storedToken.equals(refreshToken)) {
            throw new RuntimeException("Invalid refresh token");
        }

        // Generate new tokens
        String newAccessToken = jwtTokenUtil.generateAccessToken(
                account.getId(),
                account.getEmail(),
                account.getRole().getValue()
        );
        String newRefreshToken = jwtTokenUtil.generateRefreshToken(
                account.getId(),
                account.getEmail(),
                account.getRole().getValue()
        );

        // Update stored refresh token
        redisService.set(redisService.refreshTokenKey(userId), newRefreshToken, 7 * 24 * 60 * 60); // 7 days

        AccountDto.AuthResponse response = new AccountDto.AuthResponse();
        response.setAccessToken(newAccessToken);
        response.setRefreshToken(newRefreshToken);

        log.info("Refresh token successful for user: {}", userId);
        return response;
    }

    public void signOut(String refreshToken, Account account) {
        log.info("Sign out attempt for user: {}", account.getId());

        if (refreshToken == null || refreshToken.isEmpty()) {
            throw new RuntimeException("Unauthorized");
        }

        if (!jwtTokenUtil.validateToken(refreshToken)) {
            throw new RuntimeException("Invalid refresh token");
        }

        String tokenUserId = jwtTokenUtil.getUserIdFromToken(refreshToken);

        if (!tokenUserId.equals(account.getId())) {
            throw new RuntimeException("Unauthorized");
        }

        String storedToken = redisService.get(redisService.refreshTokenKey(account.getId()));
        if (storedToken == null || !storedToken.equals(refreshToken)) {
            throw new RuntimeException("Invalid refresh token");
        }

        // Delete refresh token from Redis
        redisService.delete(redisService.refreshTokenKey(account.getId()));

        log.info("Sign out successful for user: {}", account.getId());
    }



    public void verifyPin(AccountDto.VerifyPinRequest request) {
        log.info("Pin verification attempt for email: {}", request.getEmail());

        String storedCode = redisService.get(redisService.pinVerifyKey(request.getEmail()));

        if (storedCode == null) {
            throw new RuntimeException("Code expired");
        }

        if (!storedCode.equals(request.getCode())) {
            throw new RuntimeException("Invalid code");
        }

        redisService.delete(redisService.pinVerifyKey(request.getEmail()));

        log.info("Pin verified successfully for email: {}", request.getEmail());
    }

    public void forgotPassword(AccountDto.ForgotPasswordRequest request) {
        log.info("Forgot password attempt for email: {}", request.getEmail());

        Account account = accountRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("Email not found"));

        account.setPassword(passwordEncoder.encode(request.getNewPassword()));
        accountRepository.save(account);

        log.info("Password updated successfully for email: {}", request.getEmail());
    }

    private String getRandomSixDigit() {
        Random random = new Random();
        return String.format("%06d", random.nextInt(999999));
    }

    private AccountDto mapToDto(Account account) {
        AccountDto dto = new AccountDto();
        dto.setId(account.getId());
        dto.setFirstName(account.getFirstName());
        dto.setLastName(account.getLastName());
        dto.setEmail(account.getEmail());
        dto.setPhone(account.getPhone());
        dto.setGender(account.getGender() != null ? account.getGender().getValue() : null);
        dto.setAddressLine1(account.getAddressLine1());
        dto.setProvince(account.getProvince());
        dto.setDistrict(account.getDistrict());
        dto.setWard(account.getWard());
        dto.setPostalCode(account.getPostalCode());
        dto.setAddressNote(account.getAddressNote());
        dto.setDefaultAddress(account.getDefaultAddress());
        dto.setRole(account.getRole() != null ? account.getRole().getValue() : null);
        dto.setIsBlocked(account.getIsBlocked());
        dto.setCreatedAt(account.getCreatedAt());
        dto.setUpdatedAt(account.getUpdatedAt());
        return dto;
    }
}

File: be/src/services/access.service.js
import AccountModel from "../models/account.model.js";
import accessHelper from "../helpers/access.js";
import validator from "email-validator";
import bcrypt from 'bcryptjs';
import { getInfoData } from "../utils/index.js";
import redisClient from "../dbs/redisdb.js";
import jwt from "jsonwebtoken"; 
import sendEmail from "../helpers/sendEmail.js";
import dotenv from 'dotenv';
import RedisKeys from "../utils/redisKeys.js";
dotenv.config();

const getRandomSixDigit = () => {
    return Math.floor(Math.random() * (999999 - 100000 + 1)) + 100000;
}

export default class accessService {
    static signUp = async (payload) => {
        const { email, phone } = payload;

        if (!validator.validate(email)) {
            throw new Error('Invalid email');
        }

        const validEmail = await AccountModel.findOne({ email }).lean();
        if (validEmail) {
            throw new Error('Email already exists');
        }

        const validUserPhone = await AccountModel.findOne({ phone }).lean();
        if (validUserPhone) {
            throw new Error('Phone already exists');
        }

        const code = getRandomSixDigit();
        const subject = 'Email verification';
        const text = `Your code is: ${code}`;
        const info = await sendEmail(subject, text, email);
        redisClient.set(RedisKeys.userKey.pinVerify(email), code, 'EX', 60);
        return info;

    };

    static verifyEmail = async (payload) => {
        const { code, ...account } = payload;
        
        const codeToVerify = await redisClient.get(RedisKeys.userKey.pinVerify(account.email));
        console.log(`codeToVerify: ${codeToVerify}`);
        console.log(`code: ${code}`);
        console.log(code === codeToVerify);
        if (codeToVerify !== code) {
            throw new Error('Invalid code');
        }
        await redisClient.del(RedisKeys.userKey.pinVerify(account.email));

        console.log(1);
        const salt = await bcrypt.genSalt(10);
        const hashedPassword = await bcrypt.hash(account.password, salt);
        const { password, ...info } = account;
        console.log(`info: ${JSON.stringify(info)}`);
        const verifyAccount = new AccountModel({
            ...info,
            password: hashedPassword,
        });

        console.log(`verifyAccount: ${JSON.stringify(verifyAccount)}`);

        await verifyAccount.save(); 
        return verifyAccount;
    }

    static signIn = async (payload) => {
        const { email, password } = payload;

        const account = await AccountModel.findOne({ email }).lean();
        if (!account) {
            throw new Error('Invalid username or password');
        }

        const isMatch = await bcrypt.compare(password, account.password); 
        if (!isMatch) {
            throw new Error('Invalid username or password');
        }

        const fields = ['_id', 'user_name', 'email', 'role', 'isPremium', 'premiumExpired'];
        const acc = getInfoData({ fields, object: account });
        console.log(acc);

        const accessToken = accessHelper.generateToken(account, process.env.JWT_ACCESS_TOKEN, '2h');
        const refreshToken = accessHelper.generateToken(account, process.env.JWT_REFRESH_TOKEN, '7d');
        
        const redisRefreshKey = RedisKeys.userKey.refreshToken(account._id);

        await redisClient.del(redisRefreshKey);

        redisClient.set(redisRefreshKey, refreshToken);

        return { acc, accessToken, refreshToken };
    };

    static refresh = async (payload) => {
        const { refreshToken } = payload;
        console.log(`refreshToken: ${refreshToken}`);
    
        if (!refreshToken) {
            throw new Error('Unauthorized');
        }
    
        let decoded;
        try {
            decoded = jwt.verify(refreshToken, process.env.JWT_REFRESH_TOKEN);
        } catch (err) {
            if (err.name === 'TokenExpiredError') {
                throw new Error('Refresh token expired');
            }
            throw new Error('Invalid refresh token');
        }
    
        console.log(`Token expires at: ${new Date(decoded.exp * 1000).toISOString()}`);
        console.log(`Current time: ${new Date().toISOString()}`);
    
        const account = await AccountModel.findById(decoded._id).select('_id user_name email role').lean();
        if (!account) {
            throw new Error('User not found');
        }
    
        console.log(`account: ${JSON.stringify(account)}`);
    
        const redisRefreshKey = RedisKeys.userKey.refreshToken(account._id);
        const storedRefreshToken = await redisClient.get(redisRefreshKey);
        
        if (!storedRefreshToken || storedRefreshToken !== refreshToken) {
            throw new Error('Invalid refresh token');
        }
    
        const newAccessToken = accessHelper.generateToken(account, process.env.JWT_ACCESS_TOKEN, '2h');
        const newRefreshToken = accessHelper.generateToken(account, process.env.JWT_REFRESH_TOKEN, '7d');
    
        await redisClient.set(redisRefreshKey, newRefreshToken, 'EX', 7 * 24 * 60 * 60);
    
        return { newAccessToken, newRefreshToken };
    }
    
    static signOut = async (payload) => {
        const { refreshToken, account } = payload;
    
        if (!refreshToken) {
            throw new Error('Unauthorized');
        }
        console.log(process.env.JWT_REFRESH_TOKEN);
    
        const decoded = jwt.verify(refreshToken, process.env.JWT_REFRESH_TOKEN, (err, decoded) => {
            if (err) {
                throw new Error('Invalid refresh token');
            }
            return decoded;
        });
    
        if (decoded._id !== account._id) {
            throw new Error('Unauthorized');
        }
    
        const redisRefreshKey = RedisKeys.userKey.refreshToken(account._id);
        
        const storedRefreshToken = await redisClient.get(redisRefreshKey);
        if (storedRefreshToken !== refreshToken) {
            throw new Error('Invalid refresh token');
        }
    
        await redisClient.del(redisRefreshKey);
    }

    static async resetPassword(account) {
        const { email } = account;
        const code = getRandomSixDigit();
        const subject = 'Reset password';
        const text = `Your code is: ${code}`;
        const info = await sendEmail(subject, text, email);
        redisClient.set(RedisKeys.userKey.pinVerify(email), code, 'EX', 60);
        return info;
    }

    static async resentPin(account) {
        const { email } = account;
        const code = getRandomSixDigit();
        const subject = 'Resent pin';
        const text = `Your code is: ${code}`;
        const info = await sendEmail(subject, text, email);
        redisClient.set(RedisKeys.userKey.pinVerify(email), code, 'EX', 60);
        return info;
    }
    
    static verifyPin = async (payload) => {
        const { email, code } = payload;
        const codeToVerify = await redisClient.get(RedisKeys.userKey.pinVerify(email));
        if(!codeToVerify) {
            throw new Error('Code expired');
        }

        if (codeToVerify !== code) {
            throw new Error('Invalid code');
        }
        await redisClient.del(RedisKeys.userKey.pinVerify(email));
        
        return {
            message: 'Verify pin successfully'
        };
    }

    static forgotPassWord = async (payload) => {
        const { email, newPassword } = payload;
        const account = await AccountModel
        .findOne({ email })
        .select('_id');
        if (!account) {
            throw new Error('Email not found');
        }
        const salt = await bcrypt.genSalt(10);
        account.password = await bcrypt.hash(newPassword, salt);
        await account.save();
    }
}